prometheus_client/multiprocess.py

```diff
@@ -157,3 +176,43 @@ def mark_process_dead(pid, path=None):
         os.remove(f)
     for f in glob.glob(os.path.join(path, 'gauge_liveall_{0}.db'.format(pid))):
         os.remove(f)
+
+    # get associated db files with pid
+    files = glob.glob(os.path.join(path, '*_{0}.db'.format(pid)))
+    if not files:
+        return
+
+    # get merge file name
+    merge_files = []
+    for f in files:
+        file_prefix = os.path.basename(f).rsplit('_', 1)[0]
+        merge_file = os.path.join(path, '{0}_merge.db'.format(file_prefix))
+        if merge_file not in merge_files:
+            merge_files.append(merge_file)
+
+        # if not exist merge_file, create and init it
+        if not os.path.exists(merge_file):
+            MmapedDict(merge_file).close()
+
+    # do merge, here we use the same method to merge
+    metrics = MultiProcessCollector.merge(files + merge_files, accumulate=False)
```

Comment:
@brian-brazil  here I use the same method to merge metrics

Reply:
I believe this will cause deadlock: `mark_process_dead` is already holding lock and calling `merge` will try to acquire the same lock again.

Reply:
I don't believe it will -- it looks like lock acquisition via `flock` is reentrant safe, but unlocking via `flock` _isn't_, so I think this will prematurely release the lock after the `merge` call here, at least on POSIX systems. I don't know what the locking behaviour on NT will be.